daemon/logger/amqp/amqp.go

```diff
@@ -0,0 +1,419 @@
+	if ctx.Config["amqp-settings"] == "" {
+		brokerURLs = parseURL(ctx)
+	} else {
+		brokerURLs = parseJSONFile(ctx.Config["amqp-settings"])
+	}
+	logrus.Info(brokerURLs)
+	if err != nil {
+		logrus.Errorf("Invalid AMQP URL - %v", err)
+		return nil, err
+	}
+
+	currentBroker := brokerURLs[broker]
+	brokerURL := currentBroker.BrokerURL
+
+	// If the broker is being connected to using TLS then find the certificate and key files.
+	// If not then connect normally.
+	if brokerURL.Scheme == "amqps" {
+		logrus.Infof("Connecting to AMQP: %s", brokerURL)
+
+		cfg := new(tls.Config)
+		if cert, err := tls.LoadX509KeyPair(currentBroker.CertPath, currentBroker.KeyPath); err == nil {
+			cfg.Certificates = append(cfg.Certificates, cert)
+		}
+		conn, err = amqp.DialTLS(brokerURL.String(), cfg)
+		if err != nil {
+			logrus.Errorf("Could not connect to AMQP server - %v", err)
+			return nil, err
+		}
+	} else {
+		logrus.Infof("Connecting to AMQP: %s", brokerURL)
+		conn, err = amqp.Dial(brokerURL.String())
+		if err != nil {
+			logrus.Errorf("Could not connect to AMQP server - %v", err)
+			return nil, err
+		}
+	}
+
+	c, err = conn.Channel()
+	if err != nil {
+		logrus.Errorf("Could not open channel - %v", err)
+		return nil, err
+	}
+
+	if currentBroker.Confirm == true {
+		logrus.Info("Enabling publish confirmation")
+		if err := c.Confirm(false); err != nil {
+			logrus.Errorf("Could not put channel into confirm mode - %v", err)
+			return nil, err
+		}
+
+		conf = c.NotifyPublish(make(chan amqp.Confirmation, 1))
+	}
+
+	err = c.ExchangeDeclare(currentBroker.Exchange, "direct", true, false, false, false, nil)
+	if err != nil {
+		logrus.Errorf("Could not create exchange - %v", err)
+		return nil, err
+	}
+
+	_, err = c.QueueDeclare(currentBroker.Queue, true, false, false, false, nil)
+	if err != nil {
+		logrus.Errorf("Could not create queue - %v", err)
+		return nil, err
+	}
+
+	err = c.QueueBind(currentBroker.Queue, currentBroker.RoutingKey, currentBroker.Exchange, false, nil)
+	if err != nil {
+		logrus.Errorf("Could not bind queue to exchange - %v", err)
+		return nil, err
+	}
+
+	logrus.Info("Connection set up")
+	return &amqpConnection{
+		broker:     broker,
+		brokerURLs: brokerURLs,
+		conn:       conn,
+		c:          c,
+		conf:       conf,
+		err:        err,
+	}, nil
+}
+
+// If the connection fails at any point then close the current connection and
+// try to connect to the next broker in the list.
+func reconnect(s *amqpLogger) (err error) {
+	logrus.Warn("Unable to send message to AMQP broker")
+	logrus.Info("Attempting to reconnect")
+	s.Close()
+	// Move to the next broker in the list. If at the end of the
+	// list then go back to the start
+	if len(s.connection.brokerURLs) > s.connection.broker+1 {
+		s.connection.broker++
+	} else {
+		s.connection.broker = 0
+	}
+	connection, err := connect(s.ctx, s.connection.broker)
+	if err != nil {
+		logrus.Errorf("Could not reconnect: %v", err)
+		return err
```

Comment:
Yes, with the way that the reconnect function is called this would be the outcome. I need to do some more work on this and decide how many times it should go through the list of brokers before it gives up. Maybe this should be (another) configuration option.